src/generate-changelog.ts

```diff
@@ -0,0 +1,197 @@
+import {graphql} from '@octokit/graphql'
+import shell from 'shelljs'
+
+// Regexes to find the changelog contents within a PR.
+const FIXES_REGEX = /^CHANGELOG-FIXES:(.*)/gm
+const NEW_REGEX = /^CHANGELOG-NEW:(.*)/gm
+
+export interface Changelog {
+  added: string[] | undefined
+  fixed: string[] | undefined
+}
+
+interface ReleaseInfo {
+  name: string
+  version: string
+}
+
+interface GraphQLRelease {
+  name: string
+  tag: {
+    name: string
+  }
+}
+
+// Generates a changelog by parsing PRs that were newly merged into the currentVersion.
+export async function generateChangelog(
+  githubAuthToken: string,
+  currentVersion: string,
+  channel: string
+): Promise<Changelog> {
+  const graphqlWithAuth = graphql.defaults({
+    headers: {
+      authorization: `token ${githubAuthToken}`
+    }
+  })
+
+  const releases = await getReleases(graphqlWithAuth)
+
+  // Find the most recent release prior to this one, ignoring any releases that were the same version from a
+  //  prior cherrypick (e.g. v0.2022.01.01.stable_00 is still part of the same release as v0.2022.01.01.stable_01).
+  let lastReleaseVersion
+  for (const release of releases) {
+    // Only consider releases of the same type as the current channel.
+    if (
+      release.name.toLowerCase().startsWith(channel) &&
+      !release.version.startsWith(
+        currentVersion.substring(0, currentVersion.length - 1 - 2)
+      )
+    ) {
+      // Check for a release where `release.version` is less then `currentVersion`, which means `localCompare` would return `1`.
+      // This is by no means a perfect check, but should suffice because each version is of the exact same format and length.
+      if (
+        currentVersion.localeCompare(release.version, undefined /* locales */, {
+          numeric: true,
+          sensitivity: 'base'
+        }) === 1
+      ) {
+        lastReleaseVersion = release.version
+        break
+      }
+    }
+  }
+
+  if (lastReleaseVersion) {
+    // Find all the commits in the branch for the current release that aren't in the branch for the last release.
+    const currentBranch = branchFromVersion(currentVersion, channel)
+    const previousBranch = branchFromVersion(lastReleaseVersion, channel)
+
+    // Find all the commits that are in `currentBranch` but not `previousBranch`.
+    const command = shell.exec(
+      `git --no-pager log  ^${previousBranch} ${currentBranch} --pretty=format:%H`,
+      {silent: true}
+    )
+
+    const commits = command.stdout.trim().split('\n')
+    const pullRequestMetadata = await fetchPullRequestBodyFromCommits(
+      commits,
+      graphqlWithAuth
+    )
+    return parseChangelogFromPrDescriptions(pullRequestMetadata)
+  } else {
+    return Promise.reject(
+      Error('Unable to find last release prior to the given release')
+    )
```

Comment:
Rather than creating a new `Promise` here, since it's already an async function, I think it's slightly more idiomatic to throw here:

```js
throw new Error('Unable to find last release prior to the given release')
```

Also: Would it make sense to make this a negative check, rather than an `else` at the end of the block? That is, we could do something like:

```js
if (!lastReleaseVersion) {
    throw new Error(...)
}

// Rest of the code here
```

Reply:
done